### DISCUSSION TITLE
Generic with default in dataclass vs regular class

### DISCUSSION THREAD

**Discussion by maurosilber** [microsoft/pyright#6003] [Q&A] [answered]
I have an issue with type inference in instances of dataclasses, but I'm not sure if its a bug or not.

I'm defining two classes: a dataclass `A` and a regular class `B`. Both will have an attribute `func` of type `Callable[[],T]`, with a default value of `Callable[[],int]`. When creating instances and calling `func`, type inference only works for the regular class:

```python
from dataclasses import dataclass
from typing import Callable, Generic, TypeVar

T = TypeVar("T", covariant)


def f_int() -> int:
    return 1


@dataclass
class A(Generic[T]):
    func: Callable[[], T] = f_int  # type error here


class B(Generic[T]):
    def __init__(self, func: Callable[[], T] = f_int):
        self.func = func


a = A.func()      # type: T
a = A().func()  # type: Any
b = B().func()  # type: int

# It works correctly when "overriding" the default:
A(func=lambda: "s").func()  # type: str
B(func=lambda: "s").func()  # type: str
```

I understand that there is a difference between classes `A` and `B`, as `func` is also a class variable for `A`. But shouldn't it be inferred in the instance of `A` that `func` is of type `Callable[[], int]`?

**Comment by erictraut** [comment, marked as answer]
This falls into a gray area. I don't consider it a bug because it's working as designed, but I agree that the behavior appears to be inconsistent.

In general, assigning a default value to any variable or parameter that contains a type variable is fraught with problems because it can easily lead to type violations. For this reason, mypy (another Python type checker) doesn't allow this at all. It considers both of your examples above to be type violations.

Consider the expression `B[str]()`. In this case, `B` has been specialized before calling the constructor, but the supplied default argument for parameter `func` is not compatible with this specialized type. Pyright supports the default argument in this case (and does so in a type-safe manner) because it has special-case logic to verify that the default argument remains compatible even if the class has been specialized. This check is performed at the call site of the constructor. It therefore generates an error for `B[str]()` but not for `B[int]()` or for `B[float]()` (which is allowed due to covariance).

In the case of `A`, the default value is being applied to an instance variable, not to a method's default argument value. (Actually, that's not entirely true since `@dataclass` synthesizes an `__init__` method with a default argument value.) Pyright's logic for dataclass evaluation checks the type compatibility of default values when it's evaluating the field declaration. This allows pyright to report the type violation in an intelligible way to the user. For example:
```python
@dataclass
class Foo:
    a: int = ""  # Type violation reported here
```

It would be possible to move the logic for validating dataclass default field types to the call site of the dataclass constructor (e.g. for `A[str]()`). This would allow pyright to support your use case in a type-safe manner, but I think it would be a big step back in terms of usability for dataclass error reporting, so I'm reluctant to do this.

**Reply by maurosilber** [reply]
> but I think it would be a big step back in terms of usability for dataclass error reporting, so I'm reluctant to do this.

I'm not sure if I understood everything, as I'm not familiar with the inner workings of type checkers, but I agree that reporting useful errors is more important.

If this were supported, then an error would not be reported in the following case?

```python
@dataclass
class A(Generic[T]):
    x: T = 1
    y: T = ""  # error here, not compatible with T: int
```

It does work like this for an explicit `__init__`:

```python
class B(Generic[T]):
    def __init__(
        self,
        a: T = 1,
        b: T = "",  # error here, not compatible with T: int
    ):
        ...
```

Although, now that I think more about it, maybe the user intended `T` to be `int | str`.

Anyways, thank you!
